middleman_protocol/middleman_protocol/helpers.py

```diff
@@ -0,0 +1,101 @@
+from middleman_protocol.utils import escape_decode_raw_message
+from middleman_protocol.utils import escape_encode_raw_message
+
+
+def split_stream(connection: socket.socket) -> Iterator[bytes]:
+    """
+    Lowest level receiver which determines if received data is frame separator,
+    which means previously received data should be yielded,
+    or if received data is something else and receiver should continue to gather bytes.
+    """
+    assert isinstance(connection, socket.socket)
+
+    received_data = b''
+
+    while True:
+        next_bytes = connection.recv(RECEIVE_BYTES_PER_LOOP)
+
+        for next_byte in next_bytes:
+            if next_bytes == FRAME_SEPARATOR:
+                yield received_data
+                received_data = b''
+            else:
+                received_data += bytes([next_byte])
+
+
+def unescape_stream(connection: socket.socket) -> Iterator[Optional[bytes]]:
+    """
+    Middle level receiver which determines if received data has broken escaping and yields None to indicate it,
+    or reverses escaping and yields unescaped frame.
+    """
+    assert isinstance(connection, socket.socket)
+
+    for unescaped_data in split_stream(connection=connection):
+        assert isinstance(unescaped_data, bytes)
+
+        # If there is an escape character which is not a valid escape sequence,
+        # yield special value indicating a frame with broken escaping and continue the outer loop.
+        # This can be checked by comparing sum of occurrences of escape character
+        # to sum of occurrences of escape sequences. If they are equal, data is escaped correctly.
+        if (
+            unescaped_data.count(ESCAPE_CHARACTER) !=
+            sum([unescaped_data.count(escape_sequence) for escape_sequence in ESCAPE_SEQUENCES.values()])
+        ):
+            yield None
+            continue
+
+        yield escape_decode_raw_message(unescaped_data)
+
+
+def receive_frame(connection: socket.socket, public_key: bytes) -> Iterator[Optional[AbstractMiddlemanMessage]]:
+    """
+    Top level receiver which determines if received data indicates an invalid escape sequence and yields it,
+    or if received data is valid Middleman protocol message and yields instance of it.
+    """
+    assert isinstance(connection, socket.socket)
+    assert isinstance(public_key, bytes)
+
+    for received_frame in unescape_stream(connection=connection):
+        assert isinstance(received_frame, (bytes, type(None)))
+
+        # If you have just received the value indicating an invalid escape sequence,
+        # yield a special value indicating an invalid escape sequence and continue.
+        if received_frame is None:
+            yield received_frame
+
+        # Otherwise create an instance of the frame class and yield it.
+        else:
+            yield AbstractMiddlemanMessage.deserialize(
+                raw_message=received_frame,
+                public_key=public_key,
+            )
```

Comment:
What happens if this fails due to a validation error? I just noticed that it might be a design problem in our flow here. What do we expect the user of this generator to do when this raises? It will make the generator fail and lose its state after all - including any characters that it may have still have in the buffer (which is not the case now but may change if we change the implementation in the future).

@dybi What do you think?

Reply:
When it is wrapped in `try/except` and is running in a loop it works correctly - by correctly I mean that after throwing and catching exception it can still yield another data. But I am not sure about the state - I need to test it.

But for me it won't be a big deal to move it outside, as I have implemented it originally. Just some tests to update. I know @dybi wanted it to work this way, which is also good. So let me test the behavior of buffer.

Reply:
@cameel @dybi It looks like the buffer is kept, which is ok. Please check tests I have added in a9b9bf00570f096666e785a6afb0b3801dd16404.

Reply:
hm... I think it would be better if exception handling would be done in separate layer.

Reply:
I finally realized what was bugging me about this. I though that exiting from `receive_frame()` will simply kill `unescape_stream()` generator without letting it finish or clean up. But that's not what will happen. Python has a mechanism to prevent this. From the perspective of the generator, it will look as if its `yield` raised `GeneratorExit`. Generator can have a `finally` block around it which allows it to do any necessary cleanup.

This happens if you either call `close()` explicitly on the generator or you let it go out of scope so that it's garbage collected. See [GeneratorExit - How to clean up after the last yield in Python](https://www.peterbe.com/plog/generatorexit).

[Cleaning Up in a Python Generator Can Be Dangerous](https://amir.rachum.com/blog/2017/03/03/generator-cleanup/) also describes an interesting corner case.

So, anyway, in our situation the problem is that `split_stream()` may have already read a series of bytes that contains several frames from the connection. It's now going over its buffer in a loop. If `deserialize()` raises, `receive_frame()`, `unescape_frame()` and `split_stream()` will all exit, losing their state, including the buffer. If the caller then handles the exception and starts reading from the same connection again, the data that `split_stream()` has already read but not yielded yet is going to be lost.

I don't see a good solution other than going back to processing bytes one by one in `split_stream()`. Because it we read too much there's no way to return that data back to the connection.

The only other solution would be to have the generator "own" the connection and close it in its `finally` block when it's itself destroyed - so that there's no risk that anyone will continue reading from it.

Reply:
> When it is wrapped in `try/except` and is running in a loop it works correctly - by correctly I mean that after throwing and catching exception it can still yield another data. But I am not sure about the state - I need to test it.

Please test this case:
- Sender sends 240 bytes, containing 6 frames, each 40 bytes long (including separator, escaping, etc.). Frames 1, 2, 4, 5, 6 are correct. Frame 3 is malformed (cannot be deserialized).
- `split_stream()` reads 120 bytes in one go and then starts going over it in a loop.
- `split_stream()` yields the first frame. `unescape_stream()` and `receive_frame()` unescape and deserialize it. The caller gets one frame class instance.
- `split stream()` yields the second frame, `unescape_stream()` unescapes it and yields it.
- This frame is malformed so `receive_frame()` raises an exception.
- The caller handles the exception and continues reading until it exhausts the generator.

If the caller has processed 6 frames total (5 correct, 1 exception), code is correct. If it got only 5 frames total (4 correct and 1 exception), it means that the generator has lost the third frame.

Reply:
I wasn't able to test that because I have found a big issue before this. It looks like receiving more than one message at a time in `split_stream` allows it to yield only first (without exceptions along the way), and then looses rest of data. It is never yield, and when on debug the pointer jumps in a weird, unpredictable way. I think it is related with how generators are nested with each other. Lets discuss this tomorrow because it looks right now that we will have to change approach once again.

Reply:
The generator raises `StopIteration` after the malformed frame. So if we do `.deserialize()` outside generator it should work.

Reply:
> Will there be any problem with that ?

That would solve only a part of the problem. `serialize()` is not the only thing that can raise an exception. If there's an exception at any point inside the generators (e.g. while reading from the socket), we still lose the buffer.

Disregard my suggestion to read byte by byte. This won't help after all. I suggested it because then `split_stream()` can never lose a complete frame but that does not matter. Even losing a part of a frame means losing data.

There's no way around it - the socket is unusable once generator exits due to an exception. We need to:
1. **Yield broken frames**. Broken frames should be yielded, not reported with exceptions. Exception kills the generator but we want to keep reading after an invalid frame so it seems to me that the right way to model this is to represent it as just another type of data that may appear in the output. This could be simply a `(reason, raw_data)` tuple or it could be a special class.
    - We should use this mechanism also for the frames with broken escaping instead of reporting them with `None`.
    - Exceptions that do not indicate broken frames (e.g. an error when reading from the socket) should be still raised and kill the generator.
2. **Close the socket if the generator raises**. The generator should be responsible for closing the socket in a `finally` block. This does not fix it but will prevent the caller from reading more bytes from it.

Reply:
I have moved deserializing outside of the generator, however I don't understand how closing socket in `finally` block can work for us, as it will close socket even without exception. I think we should do this in `except` block.